File: src/leetcode/2026/05/Solution917.java
public class Solution917 {
    public String reverseOnlyLetters(String s) {
        char[] chars = s.toCharArray();
        int left = 0, right = chars.length - 1;

        while (left < right) {
            while (left < right && !Character.isLetter(chars[left])) {
                left++;
            }
            while (left < right && !Character.isLetter(chars[right])) {
                right--;
            }

            if (left < right) {
                char temp = chars[left];
                chars[left] = chars[right];
                chars[right] = temp;
                left++;
                right--;
            }
        }

        return new String(chars);
    }

    public static void main(String[] args) {
        Solution917 solution = new Solution917();

        String s1 = "ab-cd";
        String result1 = solution.reverseOnlyLetters(s1);
        System.out.println(result1); // Output: "dc-ba"

        String s2 = "a-bC-dEf-ghIj";
        String result2 = solution.reverseOnlyLetters(s2);
        System.out.println(result2); // Output: "j-Ih-gfE-dCba"

        String s3 = "Test1ng-Leet=code-Q!";
        String result3 = solution.reverseOnlyLetters(s3);
        System.out.println(result3); // Output: "Qedo1ct-eeLg=ntse-T!"
    }
}

File: src/leetcode/2026/05/Solution917.cpp
#include <iostream>
#include <string>
using namespace std;

class Solution917 {
public:
    string reverseOnlyLetters(string s) {
        int i = 0, j = s.size() - 1;
        while (i < j) {
            if (!isalpha(s[i])) {
                i++;
            } else if (!isalpha(s[j])) {
                j--;
            } else {
                swap(s[i], s[j]);
                i++;
                j--;
            }
        }
        return s;
    }
};

int main() {
    Solution917 solution;

    string s = "ab-cd";
    cout << solution.reverseOnlyLetters(s) << endl; // Output: "dc-ba"

    string s2 = "a-bC-dEf-ghIj";
    cout << solution.reverseOnlyLetters(s2) << endl; // Output: "j-Ih-gfE-dCba"

    string s3 = "Test1ng-Leet=code-Q!";
    cout << solution.reverseOnlyLetters(s3) << endl; // Output: "Qedo1ct-eeLg=ntse-T!"
    return 0;
}